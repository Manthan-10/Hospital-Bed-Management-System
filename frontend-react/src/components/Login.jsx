import React, { useState, useRef } from "react";
import { useNavigate } from "react-router-dom";
import Form from "react-validation/build/form";
import Input from "react-validation/build/input";
import CheckButton from "react-validation/build/button";
import AuthService from "../services/auth.service";

import "./Login.css";

/**
 * This component displays the Login form
 *
 */
const Login = () => {
  let navigate = useNavigate();

  const form = useRef();
  const checkBtn = useRef();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const [message, setMessage] = useState("");

  const required = (value) => {
    if (!value) {
      return (
        <div className="alert alert-danger" role="alert">
          This field is required!
        </div>
      );
    }
  };

  /**
   * Handling any changes while inputting username
   *
   */
  const onChangeUsername = (e) => {
    const username = e.target.value;
    setUsername(username);
  };

  /**
   * Handling any changes while inputting password
   *
   */
  const onChangePassword = (e) => {
    const password = e.target.value;
    setPassword(password);
  };

  /**
   * Handling the login
   *
   */
  const handleLogin = (e) => {
    e.preventDefault();
    setMessage("");
    setLoading(true);
    form.current.validateAll();
    if (checkBtn.current.context._errors.length === 0) {
      AuthService.login(username, password).then(
        () => {
          navigate("/hospitals/1/floor/1");
          window.location.reload();
        },
        (error) => {
          setMessage("Inavlid Username or Password");
          setLoading(false);
        }
      );
    } else {
      setLoading(false);
    }
  };

  return (
    <div className="col-md-12">
      <center>
        <h1>Hospital Bed Management System</h1>
      </center>
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />
        <Form onSubmit={handleLogin} ref={form}>
          <div className="form-group">
            <label htmlFor="username">Username</label>
            <Input
              id="username"
              type="text"
              className="form-control"
              name="username"
              value={username}
              onChange={onChangeUsername}
              validations={[required]}
            />
          </div>
          <div className="form-group">
            <label htmlFor="password">Password</label>
            <Input
              id="password"
              type="password"
              className="form-control"
              name="password"
              value={password}
              onChange={onChangePassword}
              validations={[required]}
            />
          </div>
          <div className="form-group">
            <button className="btn btn-primary btn-block" disabled={loading}>
              {loading && (
                <span className="spinner-border spinner-border-sm"></span>
              )}
              <span>Login</span>
            </button>
          </div>
          {message && (
            <div className="form-group">
              <div className="alert alert-danger" role="alert">
                {message}
              </div>
            </div>
          )}
          <CheckButton style={{ display: "none" }} ref={checkBtn} />
        </Form>
      </div>
    </div>
  );
};
export default Login;
