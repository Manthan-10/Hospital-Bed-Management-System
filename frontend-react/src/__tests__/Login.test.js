import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Login from "../components/Login";
import Adapter from "enzyme-adapter-react-16";
import * as reactRouterDom from "react-router-dom";
import renderer from "react-test-renderer";
import AuthService from "../services/auth.service";
import { render } from "@testing-library/react";
Enzyme.configure({ adapter: new Adapter() });

/**
 * Testing for UI of Login page
 * @author MN097789
 */

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  useSelector: jest.fn(),

  useNavigate: () => mockNavigate,
}));
const mockNavigateSpy = jest.spyOn(reactRouterDom, "useNavigate");

describe("Login UI Test Suite", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Login />);
  });

  test("Checks if it renders correctly", () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  //Testing if the UI is rendered correctly
  it("should check for divs", () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find("div.col-md-12").exists()).toBe(true);
    expect(wrapper.find("label").length).toEqual(2);
    expect(wrapper.find("#username").length).toEqual(1);
    expect(wrapper.find("#password").length).toEqual(1);
    expect(wrapper.find("button").length).toEqual(1);
  });

  //Testing for input change during user id
  test("Should set the username on change event", () => {
    wrapper.find("#username").simulate("change", {
      target: { value: "MN097789" },
    });
    expect(wrapper.find("#username").prop("value")).toEqual("MN097789");
  });

  //Testing for input change during password
  test("Should set the password on change event", () => {
    wrapper.find("#password").simulate("change", {
      target: { value: "12345" },
    });
    expect(wrapper.find("#password").prop("value")).toEqual("12345");
  });

  //Testing for successful login
  test("should login successfully", () => {
    const wrapper = mount(<Login />);
    wrapper.find({ username: "MN097789", password: "Manthan123" });
    const formEventMocked = { preventDefault: jest.fn() };
    wrapper.find("form").simulate("submit", formEventMocked);
    expect(formEventMocked.preventDefault).toBeCalledTimes(1);
  });
});
