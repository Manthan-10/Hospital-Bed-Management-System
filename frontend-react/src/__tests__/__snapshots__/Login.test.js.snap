// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login UI Test Suite Checks if it renders correctly 1`] = `
<div
  className="col-md-12"
>
  <center>
    <h1>
      Hospital Bed Management System
    </h1>
  </center>
  <div
    className="card card-container"
  >
    <img
      alt="profile-img"
      className="profile-img-card"
      src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="username"
        >
          Username
        </label>
        <div>
          <input
            className="form-control"
            id="username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="password"
        >
          Password
        </label>
        <div>
          <input
            className="form-control"
            id="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="form-group"
      >
        <button
          className="btn btn-primary btn-block"
          disabled={false}
        >
          <span>
            Login
          </span>
        </button>
      </div>
      
      <button
        disabled={true}
        style={
          Object {
            "display": "none",
          }
        }
      />
    </form>
  </div>
</div>
`;
